app/feature/feature-insurances/src/test/kotlin/com/hedvig/android/feature/insurances/insurancedetail/ContractDetailPresenterTest.kt

```diff
@@ -0,0 +1,340 @@
+      val presenter = ContractDetailPresenter(
+        contractId = getContractForContractIdUseCase.getValIdWithoutTerminationDate(),
+        featureManager = featureManager,
+        getContractForContractIdUseCase = getContractForContractIdUseCase,
+      )
+      presenter.test(
+        ContractDetailsUiState.Success(
+          getContractForContractIdUseCase.getInsuranceWithOutTerminationDate(),
+          true,
+        ),
+      ) {
+        assertThat(awaitItem()).isInstanceOf<ContractDetailsUiState.Success>()
+        getContractForContractIdUseCase.addErrorToResponse()
+        assertThat(awaitItem()).isInstanceOf<ContractDetailsUiState.Error>()
+        sendEvent(ContractDetailsEvent.RetryLoadingContract)
+        assertThat(awaitItem()).isInstanceOf<ContractDetailsUiState.Loading>()
+        getContractForContractIdUseCase.addInsuranceWithNoTerminationDateToResponseTurbine()
+        assertThat(awaitItem()).isInstanceOf<ContractDetailsUiState.Success>()
+      }
+    }
+
+  @Test
+  fun `with an initial error state, can retry and with a good response get success state through loading`() = runTest {
+    val getContractForContractIdUseCase = FakeGetContractForContractIdUseCase()
+    val featureManager = FakeFeatureManager(featureMap = { mapOf(Feature.TERMINATION_FLOW to true) })
+    val presenter = ContractDetailPresenter(
+      contractId = getContractForContractIdUseCase.getValIdWithoutTerminationDate(),
+      featureManager = featureManager,
+      getContractForContractIdUseCase = getContractForContractIdUseCase,
+    )
+    presenter.test(
+      ContractDetailsUiState.Error,
+    ) {
+      assertThat(awaitItem()).isInstanceOf<ContractDetailsUiState.Error>()
+      sendEvent(ContractDetailsEvent.RetryLoadingContract)
+      assertThat(awaitItem()).isInstanceOf<ContractDetailsUiState.Loading>()
+      getContractForContractIdUseCase.addInsuranceWithNoTerminationDateToResponseTurbine()
+      assertThat(awaitItem()).isInstanceOf<ContractDetailsUiState.Success>()
+    }
+  }
+
+  @Test
+  fun `with an initial error state, if a good response comes with the flow, show success state`() = runTest {
+    val getContractForContractIdUseCase = FakeGetContractForContractIdUseCase()
+    val featureManager = FakeFeatureManager(featureMap = { mapOf(Feature.TERMINATION_FLOW to true) })
+    val presenter = ContractDetailPresenter(
+      contractId = getContractForContractIdUseCase.getValIdWithoutTerminationDate(),
+      featureManager = featureManager,
+      getContractForContractIdUseCase = getContractForContractIdUseCase,
+    )
+    presenter.test(
+      ContractDetailsUiState.Error,
+    ) {
+      assertThat(awaitItem()).isInstanceOf<ContractDetailsUiState.Error>()
+      getContractForContractIdUseCase.addInsuranceWithNoTerminationDateToResponseTurbine()
+      assertThat(awaitItem()).isInstanceOf<ContractDetailsUiState.Loading>()
+      assertThat(awaitItem()).isInstanceOf<ContractDetailsUiState.Success>()
```

Comment:
I don't understand why it shows loading state in here, if we didn't emit retry event and didn't trigger dataLoadIteration. Shouldn't we just go directly from error to success here?

Reply:
Same applies here as well as far as I can tell from the presenter code. Also make sure to move the is Loading check above the `addInsuranceWithNoTerminationDateToResponseTurbine` function here too 😊